lib/streamlit/runtime/partials.py

```diff
@@ -0,0 +1,181 @@
+
+
+@overload
+def partial(
+    func: F,
+    *,
+    run_every: Optional[float] = None,
+) -> F:
+    ...
+
+
+@overload
+def partial(
+    func: None = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Callable[[F], F]:
+    ...
+
+
+def partial(
+    func: Optional[F] = None,
+    *,
+    run_every: Optional[float] = None,
+) -> Union[Callable[[F], F], F]:
+    if func is None:
+        # Support passing the params via function decorator
+        def wrapper(f: F) -> F:
+            return partial(
+                func=f,
+                run_every=run_every,
+            )
+
+        return wrapper
+    else:
+        # To make mypy type narrow Optional[F] -> F
+        non_optional_func = func
+
+    print(non_optional_func.__qualname__)
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs):
+        ctx = get_script_run_ctx()
+        if ctx is None or len(ctx.dg_stack) == 0:
+            return
+        dg_stack = ctx.dg_stack
+        active_dg = ctx.dg_stack[-1]
+
+        # TODO(lukasmasuch): Research more on what to include in the hash:
+        h = hashlib.new("md5")
+        h.update(
+            f"{non_optional_func.__module__}.{non_optional_func.__qualname__} {active_dg._get_delta_path_str()}".encode(
+                "utf-8"
+            )
+        )
+        print(non_optional_func.__qualname__)
+        partial_id = h.hexdigest()
+
+        def wrapped_partial():
+            from streamlit.runtime.scriptrunner import get_script_run_ctx
+
+            ctx = get_script_run_ctx(suppress_warning=True)
+            assert ctx is not None
+
+            ctx.dg_stack = dg_stack
+            # Set dg stack to outside state
+            print(type(ctx.dg_stack))
+            print(ctx.dg_stack)
+            ctx.current_partial_id = partial_id
+
+            result = non_optional_func(*args, **kwargs)
+
+            # TODO: always reset to None -> otherwise problems with exceptions
+            ctx.current_partial_id = None
+            return result
+
+        save_partial(partial_id, wrapped_partial)
+
+        if run_every:
+            print("Register rerun")
+            msg = ForwardMsg()
+            msg.auto_rerun.interval = run_every
+            msg.auto_rerun.partial_id = partial_id
+            ctx.enqueue(msg)
+        return wrapped_partial()
+
+    return wrap
+
+
+def load_partial(partial_id: str) -> Callable[[], Any] | None:
+    partial_storage = PartialsStorage()
+    if partial_id not in partial_storage:
+        return None
+
+    return cloudpickle.loads(partial_storage[partial_id])  # type: ignore
+
+
+def save_partial(partial_id: str, partial_function: Callable[[], Any]) -> None:
+    partial_storage = PartialsStorage()
+    partial_bytes = cloudpickle.dumps(partial_function)
+    partial_storage[partial_id] = partial_bytes
+    print(f"Size of partial {partial_id}: {sys.getsizeof(partial_bytes)}")
+
+
+class PartialsStorage(MutableMapping[str, bytes]):
```

Comment:
This is a temporary workaround to store serialized partials somewhere (currently just puts it into session state). We probably need a proper partial storage implementation, maybe it doesn't even need to be bound to a session and could be stored globally....